src/main/AllocationTable.cpp

```diff
@@ -0,0 +1,286 @@
+/*
+ *  R : A Computer Language for Statistical Data Analysis
+ *  Copyright (C) 1995, 1996  Robert Gentleman and Ross Ihaka
+ *  Copyright (C) 1998-2007   The R Development Core Team.
+ *  Copyright (C) 2008-2014  Andrew R. Runnalls.
+ *  Copyright (C) 2014 and onwards the Rho Project Authors.
+ *
+ *  Rho is not part of the R project, and bugs and other issues should
+ *  not be reported via r-bugs or other R project channels; instead refer
+ *  to the Rho website.
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, a copy is available at
+ *  https://www.R-project.org/Licenses/
+ */
+
+#include <assert.h>
+#include <stdio.h>
+
+#include <cstdint>
+#include <functional>
+#include <limits>
+
+#include "rho/AddressSanitizer.hpp"
+#include "rho/AllocationTable.hpp"
+#include "rho/GCNodeAllocator.hpp"
+
+rho::AllocationTable::AllocationTable(unsigned num_bits): m_num_bits(num_bits) {
+  m_num_buckets = 1 << m_num_bits;
+  m_capacity = m_num_buckets / 2;  // Sets max load factor to 50%.
+  m_hash_mask = m_num_buckets - 1;
+  m_buckets = new Allocation[m_num_buckets];
+}
+
+rho::AllocationTable::~AllocationTable() {
+  delete[] m_buckets;
+  m_buckets = nullptr;
+}
+
+void rho::AllocationTable::insert(void* pointer, std::size_t size) {
+  if (m_capacity == 0) {
```

Comment:
Thanks for spotting this! I now use the start and end keys to calculate the required capacity for each allocation.